File: src/packages/control-ui/src/api/enterprise_vpn_controller.ts
export const EnterpriseUserRole = {
  StandardUser: 'STANDARD_USER',
  Admin: 'ADMIN',
  Auditor: 'AUDITOR',
} as const;
export type EnterpriseUserRole = (typeof EnterpriseUserRole)[keyof typeof EnterpriseUserRole];

export interface EnterpriseTenant {
  orgId: string;
  name: string;
  virtualSubnet: string;
  maxUsers: number;
}

export interface EnterpriseUserProfile {
  username: string;
  orgId: string;
  role: EnterpriseUserRole;
  token: string;
  virtualIp: string;
  allowedRoutes: string[];
  isActive?: boolean;
}

export class EnterpriseVpnController {
  private orgs: Map<string, EnterpriseTenant> = new Map();
  private users: Map<string, EnterpriseUserProfile> = new Map();

  registerTenant(tenant: EnterpriseTenant): void {
    this.orgs.set(tenant.orgId, tenant);
  }

  registerUser(user: EnterpriseUserProfile): boolean {
    const org = this.orgs.get(user.orgId);
    if (!org) return false;

    let activeCount = 0;
    for (const u of this.users.values()) {
      if (u.orgId === user.orgId && (u.isActive ?? true)) {
        activeCount++;
      }
    }
    if (activeCount >= org.maxUsers) return false;

    this.users.set(user.token, { ...user, isActive: true });
    return true;
  }

  authenticate(token: string): EnterpriseUserProfile | null {
    const u = this.users.get(token);
    if (!u || !u.isActive) return null;
    return u;
  }

  canAccessRoute(token: string, destinationIp: string): boolean {
    const u = this.authenticate(token);
    if (!u) return false;
    if (u.role === EnterpriseUserRole.Admin) return true;

    for (const route of u.allowedRoutes) {
      if (route === '0.0.0.0/0') return true;
      if (route.includes('/')) {
        const [netStr, prefixStr] = route.split('/');
        const prefix = parseInt(prefixStr!, 10);
        if (this.matchesCidr(destinationIp, netStr!, prefix)) {
          return true;
        }
      } else if (route === destinationIp) {
        return true;
      }
    }
    return false;
  }

  revokeUser(token: string): boolean {
    const u = this.users.get(token);
    if (!u) return false;
    u.isActive = false;
    return true;
  }

  activeUserCount(): number {
    let count = 0;
    for (const u of this.users.values()) {
      if (u.isActive ?? true) count++;
    }
    return count;
  }

  private matchesCidr(ip: string, net: string, prefix: number): boolean {
    const ipVal = this.ipToNumber(ip);
    const netVal = this.ipToNumber(net);
    if (ipVal === null || netVal === null) return false;
    const mask = prefix === 0 ? 0 : (~0 << (32 - prefix)) >>> 0;
    return (ipVal & mask) === (netVal & mask);
  }

  private ipToNumber(ip: string): number | null {
    const parts = ip.split('.');
    if (parts.length !== 4) return null;
    let res = 0;
    for (const p of parts) {
      const v = parseInt(p, 10);
      if (isNaN(v) || v < 0 || v > 255) return null;
      res = (res << 8) | v;
    }
    return res >>> 0;
  }
}

File: labs/mobile/donor-archive/tunnel/EnterpriseVpnController.kt
package com.luminet.android.tunnel

enum class EnterpriseRole {
    STANDARD_USER,
    ADMIN,
    AUDITOR
}

data class EnterpriseTenant(
    val orgId: String,
    val name: String,
    val virtualSubnet: String,
    val maxUsers: Int
)

data class EnterpriseUserProfile(
    val username: String,
    val orgId: String,
    val role: EnterpriseRole,
    val token: String,
    val virtualIp: String,
    val allowedRoutes: List<String>,
    var isActive: Boolean = true
)

class EnterpriseVpnController {
    private val orgs = mutableMapOf<String, EnterpriseTenant>()
    private val users = mutableMapOf<String, EnterpriseUserProfile>() // keyed by token

    fun registerTenant(tenant: EnterpriseTenant) {
        orgs[tenant.orgId] = tenant
    }

    fun registerUser(user: EnterpriseUserProfile): Boolean {
        val org = orgs[user.orgId] ?: return false
        val activeCount = users.values.count { it.orgId == user.orgId && it.isActive }
        if (activeCount >= org.maxUsers) return false

        users[user.token] = user
        return true
    }

    fun authenticate(token: String): EnterpriseUserProfile? {
        val u = users[token] ?: return null
        return if (u.isActive) u else null
    }

    fun canAccessRoute(token: String, destinationIp: String): Boolean {
        val u = authenticate(token) ?: return false
        if (u.role == EnterpriseRole.ADMIN) return true

        for (route in u.allowedRoutes) {
            if (route == "0.0.0.0/0") return true
            if (route.contains("/")) {
                val parts = route.split("/")
                val netStr = parts[0]
                val prefixLen = parts[1].toIntOrNull() ?: 32
                if (matchesCidr(destinationIp, netStr, prefixLen)) {
                    return true
                }
            } else if (route == destinationIp) {
                return true
            }
        }
        return false
    }

    fun revokeUser(token: String): Boolean {
        val u = users[token] ?: return false
        u.isActive = false
        return true
    }

    fun activeUserCount(): Int = users.values.count { it.isActive }

    private fun matchesCidr(ip: String, net: String, prefix: Int): Boolean {
        val ipVal = ipToLong(ip) ?: return false
        val netVal = ipToLong(net) ?: return false
        val mask = if (prefix == 0) 0L else (-1L shl (32 - prefix)) and 0xFFFFFFFFL
        return (ipVal and mask) == (netVal and mask)
    }

    private fun ipToLong(ip: String): Long? {
        val parts = ip.split(".")
        if (parts.size != 4) return null
        var res = 0L
        for (p in parts) {
            val v = p.toLongOrNull() ?: return null
            if (v < 0 || v > 255) return null
            res = (res shl 8) or v
        }
        return res
    }
}
